bootstrap/src/main/java/jenkins/bootstrap/Plugin.java

```diff
@@ -0,0 +1,216 @@
+ * THE SOFTWARE.
+ */
+package jenkins.bootstrap;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.Enumeration;
+import java.util.List;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+import java.util.jar.Manifest;
+import java.util.logging.Logger;
+import java.util.zip.ZipEntry;
+
+/**
+ * Plugin that we process during the bootstrap.
+ *
+ * <p>
+ * In order to identify the core component plugin that overrides Jenkins core and put them
+ * into the classpath, the bootstrap logic needs to inspect plugins lightly.
+ * Jenkins will later do further full treatment of plugins.
+ *
+ * <p>
+ * Some of the capabilities of the full plugin manager is missing from this code. For example,
+ * the bootstrap doesn't handle {@code *.jpl} file used during the development.
+ *
+ * @author Kohsuke Kawaguchi
+ */
+class Plugin {
+    /**
+     * foo.jpi file
+     */
+    private final File archive;
+    /**
+     * Directory to explode the plugin.
+     */
+    private final File destDir;
+
+    private final Manifest manifest;
+
+    Plugin(File archive) throws IOException {
+        this.archive = archive;
+        this.destDir = getDestDir();
+        this.manifest = loadManifest();
+    }
+
+    void process(DependenciesTxt core, List<URL> jars, OverrideJournal overrides) throws IOException {
+        if (!isCoreComponentOverride() || isDisabled())
+            return;
+
+        explode();
+
+        File dependencies = new File(destDir,"META-INF/dependencies.txt");
+        boolean hasDependenciesNewerThanCore = false;
+        List<Dependency> olderThanCore = new ArrayList<>();
+        try (InputStream i = new FileInputStream(dependencies)) {
+            DependenciesTxt deps = new DependenciesTxt(i);
+            for (Dependency d : deps.dependencies) {
+                if (core.hasOlderThan(d)) {
+                    hasDependenciesNewerThanCore = true;
+                } else if (!core.hasSame(d)){
+                    olderThanCore.add(d);
+                }
+            }
+            if (!hasDependenciesNewerThanCore) {
+                // Nothing to override
+                LOGGER.info("Skipping core override plugin "+archive+" because core contains newer versions");
+                return;
+            }
+            if (olderThanCore.size() > 0)  {
+                for (Dependency d : olderThanCore) {
+                    LOGGER.warning("Skipping core override plugin " + archive + " because core contains newer versions of " + d.ga);
+                }
+                return;
+            }
+            overrides.add(core,this,deps);
+        }
+
+        File lib = new File(destDir, "WEB-INF/lib");
+        LOGGER.info("Loading core overrides from "+lib);
```

Comment:
Wait, what? Core overrides in `WEB-INF/lib`? I thought overrides were `$JENKINS_HOME/plugins/*.jpi`?

Reply:
destDir is where we exploded the override plugin